### ISSUE TITLE
Uniform python and custom mutator API

### ISSUE THREAD

**Issue by andreafioraldi** [AFLplusplus/AFLplusplus#191]
Python and custom mutators do exactly the same thing in two languages.
We should uniform the API (maintaining backward compatibility but should be easy).

I propose to define the following functions in the C mutator:
+ afl_custom_fuzz (equivalent to the old afl_custom_mutator that is used if exported and afl_custom_fuzz is not defined)
+ afl_custom_init
+ afl_custom_init_trim
+ afl_custom_trim
+ afl_custom_post_trim
+ afl_custom_pre_save (equivalent to the old afl_pre_save_handler thai is used if exported and afl_pre_save_handler is not defined)
+ afl_custom_queue_get (called when a queue entry is pulled from the queue and given to fuzz_one())

I propose to add the following method to the python mutator:
+ pre_save (does the same job of afl_custom_pre_save in the custom).
+ queue_get

In this case, a python mutator is a case of a custom mutator. We can use a single entry in the UI cause custom and python will be never used combined (and we have to check this cause they are mutually exclusive).

**Comment by domenukk** [comment]
Wrt mutually exclusive, could there be use cases where adding multiple mutators might be beneficial?

**Comment by vanhauser-thc** [comment]
they are mutually exclusive? I am not so sure about this. I thought they can be used both. and would be nice to have both.

unifying the function names etc - all for it.

**Comment by domenukk** [comment]
How about 
- supporting multiple custom mutators (multiple mutator .so files can be added with simicolons)
- refactoring the python mutators to be one custom mutator lib (or a dummy lib)
- if `AFL_PYTHON_MODULE` is set, add the python mutator lib to the path
or do you think that would be overkill?

**Comment by vanhauser-thc** [comment]
I like the idea. Although I doubt it will be a common thing people would use so not sure its something we want to put our time in? dunno.

It goes into the same direction I want to do for qemu_mode, instead (or additionally) to afl_code_start/end have afl_code_regions=0x123-0x456,0x789-0xabc,.. etc.
(but first I will do the lto instrim stuff. thats for me the absolute priority. tomorrow I will start with that)

**Comment by andreafioraldi** [comment]
I produced some examples about how will be the unified API:

C mutator: https://github.com/vanhauser-thc/AFLplusplus/blob/custom_mutator_unified_api/examples/custom_mutators/new_custom_mutator.c
Py mutator: https://github.com/vanhauser-thc/AFLplusplus/blob/custom_mutator_unified_api/examples/python_mutators/example.py

**Comment by vanhauser-thc** [comment]
afl_custom_queue_get(char* filename) <- filename is a fully qualified pathname?

afl_custom_fuzz(uint8_t *buf, size_t buf_size, uint8_t *add_buf,size_t add_buf_size,...) <- what is add_buf and where does that come from?

otherwise it looks good IMHO

**Comment by vanhauser-thc** [comment]
and I have a feature request:

afl_custom_queue_new_entry(char *filename_new_queue, char *filename_orig_queue)

This would allows for additional analysis (e.g. calling a different tool that does a different kind of coverage and saves this off to use for the custom mutator)

**Comment by andreafioraldi** [comment]
I agree, and I want to add also afl_custom_havoc_mutation that add a custom mutation to the set of the stacked mutations performed in havoc (bitflips, range insert, etc...). I'll add also afl_custom_havoc_mutation_propability that will be used to configure the probability to execute that mutations during havoc. This allow to stack mutations like in AFLSmart.

**Comment by h1994st** [comment]
Hi @andreafioraldi @domenukk @vanhauser-thc ,

I have several questions while implementing this feature (#221) and may need your help:

1. How do we use multiple custom mutators? I haven't implemented this feature, as I could not come up with a good use case. Thus, for now, the current implementation either uses the custom mutator or a Python module.

2. `fuzz` in Python module and `afl_custom_mutator` in the custom mutator library takes different arguments. Which API should we keep? Any suggestions on maintaining the backward compatibility of this API? Also, in the latest AFL++, the mutator function of the Python module is invoked in a separate stage (i.e., python_stage). Should we merge it with 

    - `fuzz` (Python) has an extra argument `add_buf`
    - `afl_custom_mutator` (the custom mutator) has two extra arguments `max_size` and `seed`

    `fuzz` (Python)
    ```python
    def fuzz(buf, add_buf):
        ret = bytearray(buf)
        ...
        return ret
    ```

    `afl_custom_mutator`/`afl_custom_fuzz` (the custom mutator)
    ```c
    size_t afl_custom_mutator(uint8_t *data, size_t size, uint8_t *mutated_out,
                              size_t max_size, unsigned int seed) {
        srand(seed);
        ...
        return mutated_size;
    }
    ```

Best

**Comment by h1994st** [comment]
Oh, I just saw the latest discussion. Thus, as shown in the latest examples, we decided to ignore the backward capability?

**Comment by vanhauser-thc** [comment]
yes we dont do backward compatability on this.
it is a 1 minute thing to upgrade an old custom mutator but would require us to keep along old code.

**Comment by domenukk** [comment]
BTW @andreafioraldi while we're redoing the custom mutator api, we might want to account for running multiple fuzz tests in parallel in the future, that would mean at least passing a `void *data` ptr to all executions, probably also an afl object of sorts containing some info about the current run (?)

**Comment by andreafioraldi** [comment]
I agree with passing an apaque pointer to the custom mutator API. In Python is a bit more problemamtic, we can skip it. @h1994st good job, when teh API is inconsitent keep the most expressive. Or you can always ask here.

**Comment by andreafioraldi** [comment]
Also, merging python and custom in a single stage is better IMO.

**Comment by h1994st** [comment]
That would be easier if we do not consider backward capability. Thank you guys for the feedback!

**Comment by h1994st** [comment]
Finished refactoring existing Python and custom mutator APIs. Pull request #221 is ready for review.

The following features are not implemented for now. I may submit a separate pull request to integrate them.

- afl_custom_queue_get(char* filename)
- afl_custom_queue_new_entry(char *filename_new_queue, char *filename_orig_queue)
- afl_custom_havoc_mutation
- afl_custom_havoc_mutation_propability

**Comment by andreafioraldi** [comment]
@vanhauser-thc what about testing the XML python mutator in travis?

**Comment by h1994st** [comment]
Hi @andreafioraldi,

You said you want to implement `afl_custom_havoc_mutation` on your own.

So, if you do not mind, I think I can work on `afl_custom_queue_get` and `afl_custom_queue_new_entry` these days.

Thanks!

**Comment by andreafioraldi** [comment]
Yes, ty, remind to always build with make ASAN_BUILD=1 and run test.sh, there was a null ptr deref!

**Comment by h1994st** [comment]
OK, got it. I will keep that in mind.

**Comment by vanhauser-thc** [comment]
> @vanhauser-thc what about testing the XML python mutator in travis?

that is a good idea. Not sure how to do this intelligently.
First LTO fixing then I will think about that

**Comment by andreafioraldi** [comment]
If I compile afl-fuzz without the Python support, there are routines that are not linked @h1994st 
```
src/afl-fuzz-mutators.c: In function 'load_custom_mutator_py':
src/afl-fuzz-mutators.c:286:7: error: 'py_functions' undeclared (first use in this function)
   if (py_functions[PY_FUNC_INIT])
       ^
src/afl-fuzz-mutators.c:286:7: note: each undeclared identifier is reported only once for each function it appears in
src/afl-fuzz-mutators.c:286:20: error: 'PY_FUNC_INIT' undeclared (first use in this function)
   if (py_functions[PY_FUNC_INIT])
                    ^
src/afl-fuzz-mutators.c:287:32: error: 'init_py' undeclared (first use in this function)
     mutator->afl_custom_init = init_py;
                                ^
src/afl-fuzz-mutators.c:291:30: error: 'fuzz_py' undeclared (first use in this function)
   mutator->afl_custom_fuzz = fuzz_py;
                              ^
src/afl-fuzz-mutators.c:293:20: error: 'PY_FUNC_PRE_SAVE' undeclared (first use in this function)
   if (py_functions[PY_FUNC_PRE_SAVE])
                    ^
src/afl-fuzz-mutators.c:294:36: error: 'pre_save_py' undeclared (first use in this function)
     mutator->afl_custom_pre_save = pre_save_py;
                                    ^
src/afl-fuzz-mutators.c:296:20: error: 'PY_FUNC_INIT_TRIM' undeclared (first use in this function)
   if (py_functions[PY_FUNC_INIT_TRIM])
                    ^
src/afl-fuzz-mutators.c:297:37: error: 'init_trim_py' undeclared (first use in this function)
     mutator->afl_custom_init_trim = init_trim_py;
                                     ^
src/afl-fuzz-mutators.c:299:20: error: 'PY_FUNC_POST_TRIM' undeclared (first use in this function)
   if (py_functions[PY_FUNC_POST_TRIM])
                    ^
src/afl-fuzz-mutators.c:300:37: error: 'post_trim_py' undeclared (first use in this function)
     mutator->afl_custom_post_trim = post_trim_py;
                                     ^
src/afl-fuzz-mutators.c:302:20: error: 'PY_FUNC_TRIM' undeclared (first use in this function)
   if (py_functions[PY_FUNC_TRIM])
                    ^
src/afl-fuzz-mutators.c:303:32: error: 'trim_py' undeclared (first use in this function)
     mutator->afl_custom_trim = trim_py;
                                ^
make: *** [afl-fuzz] Error 1
```

**Comment by andreafioraldi** [comment]
You can check if USE_PYTHON is defined using the preprocessor

**Comment by h1994st** [comment]
@andreafioraldi I see. Thank you. I should use `USE_PYTHON` to wrap up `load_custom_mutator_py`.

Will submit the pull request to fix it and the two new hooks.

**Comment by andreafioraldi** [comment]
https://github.com/vanhauser-thc/AFLplusplus/commit/172d384bf26b57beecbe084d19530ebc34a6e3fc
I changed a bit afl_custom_fuzz to avoid to allocate a huge amount of memory, @h1994st can you check it?

**Comment by h1994st** [comment]
Yes, I will check it today.

**Comment by h1994st** [comment]
> and I have a feature request:
> 
> afl_custom_queue_new_entry(char *filename_new_queue, char *filename_orig_queue)
> 
> This would allows for additional analysis (e.g. calling a different tool that does a different kind of coverage and saves this off to use for the custom mutator)

Hi @vanhauser-thc (maybe @andreafioraldi as well),

May I know why you think we need two arguments for `afl_custom_queue_new_entry`. It seems this hook will only analyze the newly added queue entry. What is the purpose of passing `filename_orig_queue`? Thank you.

**Comment by andreafioraldi** [comment]
filename_orig_queue is the testcase from which this new testcase is generated. You can use queue_cur->filename to take it.

**Comment by h1994st** [comment]
OK, I am just not very sure about the reason behind it. :(

Do you mean that, with two arguments (original and new), the custom mutator can analyze whether the mutation strategy works effectively or not? If we only take `filename_new_entry`, I think thatwe can still analyze the coverage information. Is it necessary to also include `filename_orig_entry`?

**Comment by andreafioraldi** [comment]
You may want to keep track of the generator testcase in general, is a nice thing to have it.

**Comment by h1994st** [comment]
I see. Just submitted the pull request. You can take a look. @andreafioraldi Thank you!

**Comment by andreafioraldi** [comment]
We should insert typechecks on the return values in the python mutator. If a method returns a value of a wrinf type, the fuzzer segfaults. We shout insert a descriptive error.

**Comment by domenukk** [comment]
I'm working on getting this ready for the non-global branch, however advanced custom mutators would still depend on globals with the current api. 

I suggest adding the following to the c api:
`afl_custom_init` should be of `void *` return type. 
Each function should take a `void * mutator` param, receiving the pointer returned by init. A new method `afl_custom_destroy` is called on exit to free allocations (different names like create or deinit work too). 

CC @andreafioraldi

**Comment by andreafioraldi** [comment]
Ok i approve the changes.
@h1884st do you want to join the signal group?

**Comment by domenukk** [comment]
Be aware: for the non-global AFL I'm passing `afl_state_t` to each custom function call right now as python mutators wouldn't work otherwise. 
I can change it to take the first global afl for the time being (?)

**Comment by domenukk** [comment]
@h1994st send us your phone number to gsoc@aflplus.plus

Also can you adapttthe mutator api according to
https://github.com/vanhauser-thc/AFLplusplus/issues/191#issuecomment-596450115

Thanks

**Comment by h1994st** [comment]
@domenukk 
Sure, I can try to port the custom mutator to non-global AFL.
Sent the number.

**Comment by domenukk** [comment]
Already ported it, but I had to break the api/add a pointer to the afl state.
This should be replaced with its own data pointer as discussed above.
Thanks :)

**Comment by h1994st** [comment]
Got it!
